### ISSUE TITLE
Coq's vm_compute code is broken on trunk/4.10

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#9013]
This issue is there to track the fact that Coq's vm_compute code is currently broken under trunk, due to changes to an internal API that removed the `caml_signals_are_pending` flag (that the Coq'vm code was using) in #8691.

(cc @jhjourdan @gadmm @ejgallego)

Coq-side issue: https://github.com/coq/coq/pull/10606

There is a proposal for a long-term API in #8993. (Even if that API is merged, there is the question of how Coq people could write code that is compatible with both 4.09 and 4.10.)

Having a separate issue (here) ensure that we also keep in mind the question of having a fix in time for 4.10, whether it is the long-term API proposal or some more minor short-term change.

**Comment by gasche** [comment]
A natural question: is there a "quick fix" that is possible with only a backward-compatible user-side code change (for example by exporting a tweaked definition of something as a compatibility macro) before we ask Coq to switch to the longer-term API?

**Comment by jhjourdan** [comment]
That actual Coq-side issue is coq/coq#10726

FTR, #8691 is not the only compatibility issue with Coq VM. Their issues are:
- the deprecation of `Obj.set_tag` and `Obj.truncate` : coq/coq#10602
- the removal of `caml_signals_are_pending` : coq/coq#10606
- the use of their home-made version of the `Alloc_small` macro, which is broken because of #8713.

**Comment by gasche** [comment]
Just to be clear, my idea in creating the present issue was to track "Coq is affected because `caml_signals_are_pending` is gone", not all of Coq's compatibility problems. In the case of `caml_signals_are_pending`, I am wondering whether we (the upstream) could do something to make their life easier -- so it's good to track this question here before the release. In the other cases mentioned, I suspect they will have to fend for themselves.

**Comment by jhjourdan** [comment]
My conclusion is that, in any case, because of #8713, it will not be possible to fix the runtime of OCaml to make it compatible with old versions of Coq. The solution I will propose them is the folowing:
- merge #8993 before 4.10 and use it in Coq to make it possible to handle signals in a safe manner
- use either `caml_alloc_small` or OCaml's implementation of `Alloc_small` instead of theirs.

**Comment by jhjourdan** [comment]
> Just to be clear, my idea in creating the present issue was to track "Coq is affected because caml_signals_are_pending is gone", not all of Coq's compatibility problems. In the case of caml_signals_are_pending, I am wondering whether we (the upstream) could do something to make their life easier -- so it's good to track this question here before the release. In the other cases mentioned, I suspect they will have to fend for themselves.

The fix they will need because of #8713 will anyway not be backward-compatible, so why trying to have a quick-and-dirty backward-compatibly fix that will stay forever in the OCaml codebase?

**Comment by gasche** [comment]
The two problems are independent and I would rather think of them independently, looking for the "best" solution in each case. But in any case this discussion is a bit silly: you are the one who have been doing the work so far, and it's likely that you will also end up being in charge of handling this problem, so I guess we should just go with your opinion on that. If you and @gadmm can converge on #8713 in time, great; I was just wondering whether we could have a "plan B" around that would be easier and work for the time being (~ the next few releases).

**Comment by gadmm** [comment]
@jhjourdan 
> the use of their home-made version of the Alloc_small macro, which is broken because of #8713.

Is the issue fully included in the scope of #9014 ?

**Comment by gadmm** [comment]
The actual Coq bug report and discussion is at https://github.com/coq/coq/issues/10603 :) 

@gasche 
> Even if that API is merged, there is the question of how Coq people could write code that is compatible with both 4.09 and 4.10.

This is supposed to be trivial, this is a valid and widespread use-case for preprocessor macros (the OCaml runtime makes good use of that). What is the equivalent in OCaml?

**Comment by jhjourdan** [comment]
> Is the issue fully included in the scope of #9014 ?

Not quite. Coq uses a call to `minor_collection` instead of `caml_alloc_small_dispatch` which means that, if Memprof is used, its invariants on the state of the minor heap is broken by Coq VM's allocation. Sure, it will look like it works for most users, but I would really prefer a more long-term fix, which, frankly, is not particularly hard to do. Even @silene's solution presented in https://github.com/coq/coq/issues/10603#issuecomment-537952025 would be better than this.

**Comment by gadmm** [comment]
> Even if that API is merged, there is the question of how Coq people could write code that is compatible with both 4.09 and 4.10.

After #8993 is merged, this is the only remaining question (aside from the other issues mentioned by Jacques-Henri).

After discussing with @ppedrot, the use of a preprocessor branching is the preferred solution. I think this could be made easier in general with a solution like #1798 or even just a subset of it that provides the following definition in `compatibility.h`:
```c
#define CAML_API_VERSION 410
```

**Comment by gadmm** [comment]
#1798 in fact does not do that. There, `CAML_API_VERSION` is defined by the user and means "please provide me with this API". My proposal is a `CAML_API_VERSION` parameter defined by the headers, that allows the user to apply the preprocessor idiom where one writes the branch which is destined to survive exactly as intended, and then adds the minimal deviation by branching on the available API version. This idiom allows a mechanical simplification of code at the time the user decides to drop support for an older compiler, even after people who wrote the original code are no longer available.
(The proposal for a parameter `CAML_API_VERSION` chosen by the user seems to mean that OCaml promises to keep supporting all past API versions in the future, which seems a bit untenable: it says at the same that that the API is intended to break backwards-compatibility from time to time, but that older API semantics will keep being supported indefinitely. Then better not break anything in the first place. Multicore is clearly an exceptional event from that point of view.)

**Comment by gadmm** [comment]
Reading again #1798, it seems that near the end of the discussion it is agreed that the semantics of `CAML_API_VERSION` is what I described.

I just spoke with @ppedrot who said that having such a flag would be nice, otherwise the strategy is to generate it at `configure` time by matching on `Sys.ocaml_version` which is less nice.

**Comment by xavierleroy** [comment]
`<caml/version.h>` provides macros giving the OCaml version number in various formats.
E.g. `#if OCAML_VERSION >= 41000`  will test that OCaml is 4.10 or later.  That should do the job without having to invent yet another version number.

**Comment by gadmm** [comment]
Good, this issue can be closed then. I forward this info to @ppedrot who did not seem to know this either.

**Comment by gadmm** [comment]
Downstream patch: https://github.com/coq/coq/pull/11123

**Comment by xavierleroy** [comment]
Looks like it's been fixed in Coq.

**Comment by ejgallego** [comment]
Thanks a lot to everybody, Coq + 8.10 seems to be working OK !

Something I'm not sure is fully solved or registered as a bug is the problem mentioned in  https://github.com/coq/coq/pull/9914 , cc: @silene @maximedenes ; even if that's a minor problem we should keep an issue in Coq documenting it, right?

**Comment by ejgallego** [comment]
Sorry for the noise, there is indeed an issue for this problem here https://github.com/coq/coq/issues/11170
